source/opt/dataflow.h

```diff
@@ -0,0 +1,170 @@
+  enum class VisitResult {
+    // The analysis result for this instruction has changed.
+    // This means that any instructions that depend on it (its successors) must
+    // be recomputed.
+    kResultChanged,
+    // The analysis result for this instruction has not changed.
+    // When all visit operations return |kResultFixed|, the analysis has reached
+    // a fixed point (converged).
+    kResultFixed,
+  };
+
+  virtual ~DataFlowAnalysis() {}
+
+  // Run this analysis on a given function.
+  // For analyses which work interprocedurally, |function| may be ignored.
+  void Run(Function* function);
+
+ protected:
+  DataFlowAnalysis(IRContext& context)
+      : on_worklist_(),
+        worklist_(WorklistComparator(*this)),
+        context_(context) {}
+
+  // Initialize the worklist for a given function.
+  // |is_first_iteration| is true on the first call to |Run| and false
+  // afterwards. All subsequent runs are only necessary to check if the analysis
+  // has converged; if |EnqueueSuccessors| is complete, |InitializeWorklist|
+  // should do nothing after the first iteration.
+  virtual void InitializeWorklist(Function* function,
+                                  bool is_first_iteration) = 0;
+
+  // Enqueues the successors (instructions which use the analysis result) of
+  // |inst|. This is not required to be complete, but convergence is faster when
+  // it is. This is called whenever |Visit| returns |kResultChanged|.
+  virtual void EnqueueSuccessors(Instruction* inst) = 0;
+
+  // Visits the given instruction, recomputing the analysis result. This is
+  // called once per instruction queued in |InitializeWorklist| and afterward
+  // when a predecessor is changed, through |EnqueueSuccessors|.
+  virtual VisitResult Visit(Instruction* inst) = 0;
+
+  // Returns true if |first| should be visited after |second|.
+  // This is used as a comparator for the worklist.
+  virtual bool InstructionShouldBeVisitedAfter(Instruction* first,
+                                               Instruction* second) = 0;
+
+ protected:
+  // Enqueues the given instruction to be visited. Ignored if already in the
+  // worklist.
+  bool Enqueue(Instruction* inst);
+
+ private:
+  // Runs one pass, calling |InitializeWorklist| and then iterating through the
+  // worklist until all fixed.
+  VisitResult RunOnce(Function* function, bool is_first_iteration);
+
+  friend class WorklistComparator;
+  class WorklistComparator {
+    DataFlowAnalysis& analysis_;
+
+   public:
+    WorklistComparator(DataFlowAnalysis& analysis) : analysis_(analysis) {}
+
+    bool operator()(Instruction* first, Instruction* second) {
+      return analysis_.InstructionShouldBeVisitedAfter(first, second);
+    }
+  };
+
+ protected:
+  std::unordered_map<Instruction*, bool> on_worklist_;
+  std::priority_queue<Instruction*, std::vector<Instruction*>,
+                      WorklistComparator>
+      worklist_;
```

Comment:
Why are these exposed to derived classes? Since you already have methods like `Enqueue`, `InitializeSuccessors`, how else can these be used by subclasses?